### DISCUSSION TITLE
Neurokit and ProMAC R-peak detectors

### DISCUSSION THREAD

**Discussion by levsilva** [neuropsychology/NeuroKit#599] [Q&A] [answered]
Hello!

I am using Neurokit2 for ECG delineation and I will use the "neurokit" and "promac" methods for R-peak detection. I had a look on the documentation page, the source code and paper but I could not find references for these two methods. Even if there is not specific paper for these methods, may you provide some references for where did you inspire on to define these two methods?

I take the opportunity to thank all the team for your hard work with Neurokit2.

Best regards,
Luiz

**Comment by zen-juen** [comment, marked as answer]
Thanks for reaching out! 😄 @DominiqueMakowski I think you added the promac method, though quite a long time back

**Reply by levsilva** [reply]
Hi @zen-juen, thanks for your reply. While we are waiting for @DominiqueMakowski's comment about "promac", would you know anything about the "neurokit" peak detector?

Thank you!

**Reply by DominiqueMakowski** [reply]
Please see https://github.com/neuropsychology/NeuroKit/issues/476 for the 'neurokit' method. As for the 'promac', I have no memory but it's possible that it is documented somewhere in the ecg-detectors or biosppy github repositories

**Reply by levsilva** [reply]
Thank you @DominiqueMakowski for your reply.

Considering the results of the comparison of peak detectors you have in the Neurokit2 doc page, the 'neurokit' algorithm definitely deserves a publication. Are you still considering publishing that study?

About the 'promac' method, I could not find any reference for it either in ecg-detectors or biossppy. I also tried a google search, without success. The idea behind it is very interesting and it would be useful having the possibility for controlling its parameters. For the dataset I am workin with (from children), the method 'rodrigues2021' often crashes and prevent the whole process to continue. When I removed it from the pipeline, I got better results. Also, I tested different SD for the gaussian PDF used in the convolutions and found that 90 ms was better than 100 ms in my case. But I had to change the local source code. Maybe that is something I could contribute but I have never worked with collaborative coding... 🙈

**Reply by DominiqueMakowski** [reply]
## About 'neurokit' 

> Are you still considering publishing that study?

A validation / benchmarking study would be great but we never found the time... but recently @cbrnr did some cool work [on that front](https://sleepecg.readthedocs.io/en/stable/heartbeat_detection.html), but I don't know whether he plans to publish it somehow.  

## About promac

I checked and it was actually me that added it 🙈 see https://github.com/neuropsychology/NeuroKit/issues/222
I think "promac" stands for something like "Probability... Maximum? Convolution?"... I don't remember :/ note to myself to always document more my exploratory stuff. The idea was to leverage the power of many methods but it never got tested and validated properly... 

Feel free to try to improve it, and use neurokit as your playground to get into collaborative coding :) 

what you could try perhaps is to put the methods that sometimes fail 'rodrigues2021' (btw, why does it fail? maybe it's fixable) in a try/except block so that a failure doesn't compromise the whole thing, and expose some of the variables (i.e., they are not hardcoded but as arguments that can be modified). You can try to make a pull request and we'll help you out

**Reply by cbrnr** [reply]
> A validation / benchmarking study would be great but we never found the time... but recently @cbrnr did some cool work on that front, but I don't know whether he plans to publish it somehow.

Tagging @hofaflo who implemented all of this. I was thinking about publishing our comparisons at some point, but of course I haven't found the time yet. Maybe something we might collaborate on?

**Reply by DominiqueMakowski** [reply]
definitely :)

**Reply by cbrnr** [reply]
Great! Then I'll try to come up with a plan and we should discuss the next steps.

**Reply by levsilva** [reply]
@DominiqueMakowski If I understand it right, the idea of ProMAC (Probabilistic Methods-Agreement via Convolution) is original for a peak detector. Maybe you and @cbrnr could consider including it in the validation/benchmarking study.

> what you could try perhaps is to put the methods that sometimes fail 'rodrigues2021' (btw, why does it fail? maybe it's fixable) in a try/except block so that a failure doesn't compromise the whole thing, and expose some of the variables (i.e., they are not hardcoded but as arguments that can be modified). You can try to make a pull request and we'll help you out

The error with 'rodrigues2021' is an IndexError:
```
File "...", line 993, in _ecg_findpeaks_rodrigues
    if processed_ecg[i] > Rpeakamp:
IndexError: index 4988 is out of bounds for axis 0 with size 4988
```

I noticed that sometimes (but much less often then 'rodrigues'), 'engzee' method also crashes with a similar error:

```
File "...", line 777, in _ecg_findpeaks_engzee
    r_peaks.pop(0)  # removing the 1st detection as it 1st needs the QRS complex amplitude for the threshold
IndexError: pop from empty list
```

So, maybe the best approach would be individually isolating each method in separated try-except, so that we can still benefit from the other successful methods. Does it make sense?

I will start playing with github and Neurokit2 repo to work on this features for 'promac'.

**Reply by DominiqueMakowski** [reply]
> Does it make sense?

Yes exactly. I also ping @TiagoTostas who implemented "rodrigues" in case he can think of a fix.

Also, @zen-juen @anshu-97, engzee can probably be fixed by checking before the pop `if len(r_peaks) == 0:` then basically return an empty list (if means no peak were detected).

> I will start playing with github and Neurokit2 repo to work on this features for 'promac'.

Good luck :) and let us know if there's anthing

**Reply by levsilva** [reply]
Great, thank you!

I have already started working on it and I have some ideas on changing "ecg_findpeaks" and "ecg_peaks", which includes the documentation. Should we discuss the ideas here (or in any other place) before pulling a request or do you usually review everything only after pulling?

**Reply by DominiqueMakowski** [reply]
I'd say the simplest way would be for you to open a pull request as fast as possible, and then add commits to it "little by little" so we can review and discuss bits by bits, rather than all at once. Essentially, once you open a pull request from your fork to the main one (don't forget to put yourself on the dev branch and to make a PR to the dev branch - but don't worry too much either it's something we can change later), we can continue working on it and add commits (see for instance the opened pull request for the dev -> master PR #579)
